atc-router.go

```diff
@@ -113,3 +115,95 @@ func (r *Router) GetFields() ([]string, error) {
 
 	return flds, nil
 }
+
+const (
+	defaultNumFields    = 100
+	defaultMaxFieldSize = 100
+	errBufsize          = 500
+)
+
+var (
+	fieldsBuf = []C.uchar{}
+	fieldsLen = C.ulong(defaultMaxFieldSize * defaultNumFields)
+	fieldsNum = C.ulong(defaultNumFields)
+	errorBuf  = [errBufsize]C.uchar{}
+)
+
+type ValidationResult struct {
+	fields    []string
+	operators uint64
+	errorMsg  string
+}
+
+func ValidateExpression(s Schema, atc string) *ValidationResult {
+	atcC := unsafe.Pointer(C.CString(atc))
+	defer C.free(atcC)
+
+	operators := C.ulong(0)
+	errorBufLen := C.ulong(errBufsize)
+
+loop:
+	for {
+		if len(fieldsBuf) < int(fieldsLen) {
+			fieldsBuf = make([]C.uchar, fieldsLen)
+		}
+
+		switch C.expression_validate(
+			(*C.uchar)(atcC), s.s,
+			(*C.uchar)(&fieldsBuf[0]), &fieldsLen, &fieldsNum,
+			&operators,
+			&errorBuf[0], &errorBufLen) {
+		case 0:
+			break loop
+		case 1:
+			return &ValidationResult{
+				errorMsg: C.GoStringN(
+					(*C.char)(unsafe.Pointer(&errorBuf[0])),
+					C.int(errorBufLen),
+				),
+			}
+		case 2:
+			continue
+		}
+	}
+	return &ValidationResult{
+		fields:    splitByNulls(fieldsBuf, int(fieldsNum)),
+		operators: uint64(operators),
+	}
+}
+
+const (
+	BinaryOperatorFlags_EQUALS = 1 << iota
+	BinaryOperatorFlags_NOT_EQUALS
+	BinaryOperatorFlags_REGEX
+	BinaryOperatorFlags_PREFIX
+	BinaryOperatorFlags_POSTFIX
+	BinaryOperatorFlags_GREATER
+	BinaryOperatorFlags_GREATER_OR_EQUAL
+	BinaryOperatorFlags_LESS
+	BinaryOperatorFlags_LESS_OR_EQUAL
+	BinaryOperatorFlags_IN
+	BinaryOperatorFlags_NOT_IN
+	BinaryOperatorFlags_CONTAINS
+)
+
+func splitByNulls(b []C.uchar, maxN int) []string {
+	out := make([]string, maxN)
+	pos := 0
+	for i := range out {
+		fieldLen := slices.Index(b[pos:], C.uchar(0))
+		if fieldLen <= 0 {
+			break
+		}
+
+		str := C.GoStringN((*C.char)(unsafe.Pointer(&b[pos])), C.int(fieldLen))
+
+		out[i] = str
+		pos += fieldLen + 1
+		if pos >= len(b) {
+			break
+		}
+	}
+	slices.Sort(out)
```

Comment:
first because the tests were failing half the time.  i thought about sorting in the tests but i think if the sets are to be cached, it's safer to have them in some consistent order.

of course, an alternative is to use a set (actually `map[string]struct{}`), but i would be surprised if creating and filling one isn't way heavier than sorting a slice.